server/src/main/java/org/elasticsearch/common/util/ReleasableByteArray.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.util.BigArrays.indexIsInt;
+
+public class ReleasableByteArray implements ByteArray {
+
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(ReleasableByteArray.class);
+
+    private final ReleasableBytesReference ref;
+
+    ReleasableByteArray(StreamInput in) throws IOException {
+        this.ref = in.readReleasableBytesReference();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBytesReference(ref);
+    }
+
+    @Override
+    public long size() {
+        return ref.length() / Byte.BYTES;
+    }
+
+    @Override
+    public byte get(long index) {
+        assert indexIsInt(index);
+        return ref.get((int) index);
+    }
+
+    @Override
+    public boolean get(long index, int len, BytesRef ref) {
+        assert indexIsInt(index);
+        BytesReference sliced = this.ref.slice((int) index, len);
+        if (sliced.length() != 0) {
+            ref.offset = sliced.arrayOffset();
+            ref.length = sliced.length();
+            ref.bytes = sliced.array();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public byte set(long index, byte value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void set(long index, byte[] buf, int offset, int len) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void fill(long fromIndex, long toIndex, byte value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean hasArray() {
+        return ref.hasArray();
+    }
+
+    @Override
+    public byte[] array() {
+        return ref.array();
```

Comment:
This seems dangerous. There's a weird assumption in the arrays returned by this method in so far that we assume that valid data starts at index `0`.
This is not guaranteed for arrays resulting from `in.readReleasableBytesReference();` (as a matter of fact, it's rather unlikely.

If we want to support this, which seems nice, we'd have to enhance the interface to return an `arrayOffset` just like we do with `org.elasticsearch.common.bytes.BytesReference#arrayOffset`.

Reply:
Thanks for spotting this! Maybe for now shouldn't support this? I can add a comment here and throw unsupported operation exception. Then if/when this really gets used the interface can be improved and supported in a safe manner.

Reply:
Sounds good, lets make it not supported here for now then :)
This was used as an optimization for the path `ByteArray` -> `BytesReference` which is quite nice. I can open a suggested fix PR once this is in :)